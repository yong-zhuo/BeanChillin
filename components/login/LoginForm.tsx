'use client'

import FormInput from "../form/FormInput";
import { loginFields } from "@/constants/formFields";
import Header from "../form/Header";
import Link from "next/link";
import FormButton from "../form/FormButton";
import { useState } from "react";
import userAuthentication from "../users/UserAuthentication";
import { useRouter } from "next/navigation";
import { FormEvent } from 'react';
const fields = loginFields;

interface LoginFieldsState {
  [key: string]: string;
}
let fieldsState: LoginFieldsState = {};
fields.forEach(field => fieldsState[field.id] = '');

const LoginForm = () => {
  const router = useRouter();
  const [loginState, setLoginState] = useState(fieldsState);
  const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
    setLoginState({ ...loginState, [e.target.name]: e.target.value });
  };

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>): Promise<void> => {
    e.preventDefault();

    try {
      const formData = new FormData(e.currentTarget);
      const email = formData.get('Email') as string;
      const password = formData.get('Password') as string;
      const isAuth = await userAuthentication(email, password);
      if (!isAuth) {
        throw new Error("Authentication failed");
      }
      router.push("/home");
    } catch (e) {
      console.log("unsuccessful login. Please try again!", e);
    }
  };

  return (
    <div className="h-full min-h-screen w-full bg-[url('/patterns/pattern-light.png')]">
      <Header
        heading="Login to your account"
        paragraph="Don't have an account yet? "
        linkName="Sign up now"
        linkUrl="/register"
      />

      <form onSubmit={handleSubmit} className="mb-4 px-40 pb-8 pt-6">
        {fields.map((field) => (
          <FormInput
            key={field.id}
            handleChange={handleChange}
            value={loginState[field.id]}
            labelText={field.labelText}
            id={field.id}
            name={field.name}
            type={field.type}
            isRequired={field.isRequired}
            placeholder={field.placeholder}
          />
        ))}

        <div className="flex justify-end text-sm -mt-5">
          <Link
            href="/forgot-password"
            className="font-medium text-primary hover:underline"
          >
            Forgot your Password?
          </Link>
        </div>

        <FormButton text="Sign In" action="submit" />
      </form>
    </div>
  );
};

export default LoginForm;
