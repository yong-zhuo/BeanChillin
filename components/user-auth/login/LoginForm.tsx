"use client";

import FormInput from "@/components/common-ui/form/FormInput";
import { loginFields } from "@/constants/formFields";
import Header from "@/components/common-ui/form/Header";
import Link from "next/link";
import Button from "@/components/common-ui/button/Button";
import { useRouter } from "next/navigation";
import { fieldState } from "@/types/formFieldsState";
import GoogleButton from "../common-auth-ui/GoogleButton";
import Divider from "@/components/common-ui/misc/Divider";
import { useSession, signIn } from "next-auth/react";
import { SubmitHandler, useForm } from "react-hook-form";
import { zodResolver } from "@hookform/resolvers/zod";
import { loginSchema, type login } from "@/lib/schemas/loginSchema";
import { useToast } from "@/components/common-ui/shadcn-ui/toast/use-toast";
import { ToastAction } from "@/components/common-ui/shadcn-ui/toast/toast";
import { useState } from "react";


//login fields to be mapped
const fields = loginFields;
let fieldsState: fieldState = {};
fields.forEach((field) => (fieldsState[field.id] = ""));

const LoginForm = () => {
  const [isLoading, setIsLoading] = useState(false);

  //zod validation for login
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<login>({ resolver: zodResolver(loginSchema) });

  //to use shadcn-ui toast
  const { toast } = useToast();

  //reroute user to home if session detected
  const router = useRouter();
  const { data: session } = useSession();
  if (session) {
    router.push("/home");
  }

  //if login credentials match, send users to home
  //else throw an error and display toast
  const onSubmit: SubmitHandler<login> = async (data) => {
    setIsLoading(true);
    try {
      const result = await signIn("credentials", {
        email: data.email,
        password: data.password,
        redirect: false,
        callbackUrl: "/home",
      });
      if (!result?.ok) {
        throw new Error(result?.error?.toString());
      }
    } catch (e) {
      toast({
        variant: "destructive",
        title: "Invalid email and password",
        description:
          "Please try again or register a new account if you have not done so.",
        action: <ToastAction altText="Dismiss">Dismiss</ToastAction>,
      });
      setIsLoading(false);
    }
  };

  return (
    <>
      <Header
        heading="Login to your account"
        paragraph="Don't have an account yet? "
        linkName="Sign up now"
        linkUrl="/register"
        logo
      />

      <form
        onSubmit={handleSubmit(onSubmit)}
        className="m-5 lg:mb-5 lg:px-32 lg:pb-2 lg:pt-6 "
      >
        {fields.map((field) => (
          <FormInput
            key={field.id}
            labelText={field.labelText}
            id={field.id}
            type={field.type}
            placeholder={field.placeholder}
            register={register}
            name={field.name as keyof login}
            error={errors[field.name as keyof login]}
          />
        ))}

        <div className="-mt-5 flex justify-end text-sm sm:text-xs">
          <Link
            href="/forget-password"
            className="font-medium text-pri hover:underline sm:text-base md:text-sm lg:text-sm "
          >
            Forgot your Password?
          </Link>
        </div>
        <div className="mx-3 my-3 px-5 min-w-1/2">
          <Button
            text="Sign In"
            action="submit"
            addClass="text-white bg-pri shadow-md hover:bg-slate-400 transition ease-in-out duration-300 delay-50 hover:-translate-y-1 hover:scale-110"
            state={isLoading}
            height={20}
            width={20}
          />
        </div>

        <Divider text="or" borderColor="border-pri" textColor="pri" />
      </form>
      <div className="lg:-pt-6 m-5 lg:-mt-10 lg:mb-4 lg:px-32 lg:pb-8">
        <div className="-py-2 mx-3 px-5">
          <GoogleButton />
        </div>
      </div>
    </>
  );
};

export default LoginForm;
