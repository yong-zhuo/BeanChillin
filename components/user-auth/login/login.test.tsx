/* eslint-disable testing-library/no-wait-for-side-effects */
import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import LoginForm from "./LoginForm";
import "@testing-library/jest-dom/";
import { useRouter } from "next/navigation";
import { SessionProvider, signIn } from "next-auth/react";
import { Session } from "next-auth";

let mockedSession: Session | null = null;

jest.mock("next/navigation", () => ({
  useRouter: jest.fn().mockReturnValue({
    push: jest.fn(),
  }),
}));
jest.mock("next-auth/next", () => ({
  getServerSession: jest.fn(() => Promise.resolve(mockedSession)),
}));

describe("LoginForm", () => {
  test("snapshot matches", async () => {
    (useRouter as jest.Mock).mockReturnValue({
      push: jest.fn(),
    });
    const { asFragment } = render(
      <SessionProvider session={mockedSession}>
        <LoginForm />
      </SessionProvider>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});



describe("Login Form", () => {
  test("displays error message when given empty fields", async () => {
    render(
      <SessionProvider session={mockedSession}>
        <LoginForm />
      </SessionProvider>,
    );


    // Simulate form submission without providing any input
    fireEvent.click(screen.getByRole("button", { name: "Sign In" }));

    //error messages expected
    await waitFor(() => {
      expect(screen.getByText('Email is required')).toBeInTheDocument();
      const passwordErrors = screen.getAllByText('At least 8 characters long');
      // eslint-disable-next-line testing-library/no-wait-for-multiple-assertions
      expect(passwordErrors).toHaveLength(1);
    });
  });
});