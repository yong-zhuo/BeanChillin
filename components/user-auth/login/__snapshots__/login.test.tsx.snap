// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm snapshot matches 1`] = `
<DocumentFragment>
  <div
    class="mb-5"
  >
    <div
      class="flex justify-center"
    >
      <div
        class="flex justify-center p-10"
      >
        <img
          alt="Logo"
          data-nimg="1"
          decoding="async"
          height="20"
          loading="lazy"
          src="/logo/logo.svg"
          style="color: transparent;"
          width="230"
        />
      </div>
    </div>
    <h2
      class="text-center text-nowrap text-2xl md:text-3xl font-extrabold text-gray-700"
    >
      Login to your account
    </h2>
    <p
      class="mt-2 text-center text-sm md:text-lg text-gray-500"
    >
      Don't have an account yet?  
      <a
        class="font-medium text-pri hover:underline"
        href="/register"
      >
        Sign up now
      </a>
    </p>
  </div>
  <form
    class="m-5 lg:mb-5 lg:px-32 lg:pb-2 lg:pt-6 "
  >
    <div
      class="flex justify-between"
    >
      <label
        class="text-md text-black-700 mb-2 block font-bold sm:text-base md:text-sm"
        for="email"
      >
        Email 
      </label>
    </div>
    <div
      class="mb-7"
    >
      <label
        class="hidden"
        for="email"
      >
        Email
      </label>
      <input
        class="shadow rounded-md appearance-none relative block w-full px-3 py-2 border placeholder-gray-400 text-gray-900 focus:outline-none focus:ring-pri focus:border-pri focus:z-10 sm:text-sm "
        data-testid="email"
        name="email"
        placeholder="e.g. bean@chillin.com"
        type="email"
      />
    </div>
    <div
      class="flex justify-between"
    >
      <label
        class="text-md text-black-700 mb-2 block font-bold sm:text-base md:text-sm"
        for="password"
      >
        Password 
      </label>
    </div>
    <div
      class="mb-7"
    >
      <label
        class="hidden"
        for="password"
      >
        Password
      </label>
      <input
        class="shadow rounded-md appearance-none relative block w-full px-3 py-2 border placeholder-gray-400 text-gray-900 focus:outline-none focus:ring-pri focus:border-pri focus:z-10 sm:text-sm "
        data-testid="password"
        name="password"
        placeholder="••••••••"
        type="password"
      />
    </div>
    <div
      class="-mt-5 flex justify-end text-sm sm:text-xs"
    >
      <a
        class="font-medium text-pri hover:underline sm:text-base md:text-sm lg:text-sm "
        href="/forget-password"
      >
        Forgot your Password?
      </a>
    </div>
    <div
      class="mx-3 my-3 px-5 min-w-1/2"
    >
      <button
        class="rounded-md w-full group relative flex justify-center border border-transparent text-sm font-medium py-2 px-4 mt-9 text-white bg-pri shadow-md hover:bg-slate-400 transition ease-in-out duration-300 delay-50 hover:-translate-y-1 hover:scale-110"
        data-testid="Sign In"
        type="submit"
      >
        <div
          class="flex flex-row items-center gap-1 overflow-hidden text-ellipsis whitespace-nowrap"
        >
          Sign In
        </div>
      </button>
    </div>
    <div
      class="relative my-8 flex items-center"
    >
      <div
        class="flex-grow border-t border-pri"
      />
      <span
        class="mx-4 text-pri"
      >
        or
      </span>
      <div
        class="flex-grow border-t border-pri"
      />
    </div>
  </form>
  <div
    class="lg:-pt-6 m-5 lg:-mt-10 lg:mb-4 lg:px-32 lg:pb-8"
  >
    <div
      class="-py-2 mx-3 px-5"
    >
      <button
        class="rounded-md w-full group relative flex justify-center border border-transparent text-sm font-medium py-2 px-4 mt-9 bg-white hover:bg-gray-300 delay-50 gap-2 shadow-md transition ease-in-out duration-300 hover:-translate-y-1 hover:scale-110"
        data-testid="Sign In with Google"
      >
        <img
          alt="Google"
          data-nimg="1"
          decoding="async"
          height="20"
          loading="lazy"
          src="/_next/image?url=%2Fmisc%2FGoogle.png&w=48&q=75"
          srcset="/_next/image?url=%2Fmisc%2FGoogle.png&w=32&q=75 1x, /_next/image?url=%2Fmisc%2FGoogle.png&w=48&q=75 2x"
          style="color: transparent;"
          width="20"
        />
        <div
          class="flex flex-row items-center gap-1 overflow-hidden text-ellipsis whitespace-nowrap"
        >
          Sign In with Google
        </div>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
